### DISCUSSION TITLE
Overfitting on ppocr_v4 recognition

### DISCUSSION THREAD

**Discussion by kevinzezel** [PaddlePaddle/PaddleOCR#13631] [Q&A]
Hi @GreatV

We managed to finetune the en_PP-OCRv4_rec model in English for a dataset with 7 fixed characters in each image. We achieve 95% accuracy in validation base. Is a dataset of car license plates that have 7 fixed characters.

However, now I need to train on another dataset that may have 3 or 4 characters (Motorcycle dataset). But the training accuracy reaches 98% in the first epoch and the validation metric does not exceed 60%. Would you have any suggestions on what I can change in en_PP-OCRv4_rec.yaml ? We tried several things, but without success. 

```
Global:
  debug: false
  use_gpu: true
  epoch_num: 400
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ./output/rec_ppocr_v4_Train_v3.13
  save_epoch_step: 10
  eval_batch_step:
  - 0
  - 200
  cal_metric_during_train: true
  pretrained_model: null
  checkpoints: null
  pretrained_model: /root/src/pretrain_models/en_PP-OCRv4_rec_train/best_accuracy.pdparams
  # checkpoints: /root/src/pretrain_models/en_PP-OCRv4_rec_train/best_accuracy
  save_inference_dir: null
  use_visualdl: false
  infer_img: /root/paddle/PaddleOCR/doc/imgs_words/ch/word_1.jpg
  character_dict_path: /root/paddle/PaddleOCR/ppocr/utils/en_dict.txt
  max_text_length: 25
  infer_mode: false
  use_space_char: true
  distributed: true
  save_res_path: ./output/rec/predicts_ppocrv4_Train_v3.13.txt
Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Const
    learning_rate: 0.0000002
    warmup_epoch: 0
  regularizer:
    name: L2
    factor: 3.0e-02
Architecture:
  model_type: rec
  algorithm: SVTR_LCNet
  Transform: null
  Backbone:
    name: PPLCNetV3
    scale: 0.85
  Head:
    name: MultiHead
    head_list:
    - CTCHead:
        Neck:
          name: svtr
          dims: 120
          depth: 2
          hidden_dims: 120
          kernel_size:
          - 1
          - 3
          use_guide: true
        Head:
          fc_decay: 1.0e-05
    - NRTRHead:
        nrtr_dim: 40
        max_text_length: 25
Loss:
  name: MultiLoss
  loss_config_list:
  - CTCLoss: null
  - NRTRLoss: null
PostProcess:
  name: CTCLabelDecode
Metric:
  name: RecMetric
  main_indicator: acc
  ignore_space: false
Train:
  dataset:
    name: MultiScaleDataSet
    ds_width: false
    data_dir: /root/train_dataset/v4.1/imgs/Train
    ext_op_transform_idx: 1
    label_file_list:
    - /root/train_dataset/v4.1/imgs/Train.txt
    transforms:
    - DecodeImage:
        # img_mode: BGR
        img_mode: BGR
        channel_first: false
    # - RecConAug:
    #     prob: 0.5
    #     ext_data_num: 2
    #     image_shape:
    #     - 48
    #     - 320
    #     - 3
    #     max_text_length: 25
    - RecAug: null
    - MultiLabelEncode:
        gtc_encode: NRTRLabelEncode
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_gtc
        - length
        - valid_ratio
  sampler:
    name: MultiScaleSampler
    scales:
    - - 320
      - 32
    - - 320
      - 48
    - - 320
      - 64
    first_bs: 96
    fix_bs: false
    divided_factor:
    - 8
    - 16
    is_training: true
  loader:
    shuffle: true
    batch_size_per_card: 96
    drop_last: False
    num_workers: 1
Eval:
  dataset:
    name: SimpleDataSet
    data_dir: /root/train_dataset/v3-kevin/v3/imgs/cropped_3chars/Val
    label_file_list:
    - /root/train_dataset/v3-kevin/v3/imgs/cropped_3chars/Val.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - MultiLabelEncode:
        gtc_encode: NRTRLabelEncode
    - RecResizeImg:
        image_shape:
        - 3
        - 48
        - 320
    - KeepKeys:
        keep_keys:
        - image
        - label_ctc
        - label_gtc
        - length
        - valid_ratio
  loader:
    shuffle: false
    drop_last: false
    batch_size_per_card: 128
    num_workers: 1
profiler_options: null
```

We tried to reduce NRTRHead nrtr_dim to 40, but it has been shown to be ineffective. We tried changing the PPLCNetV3 scale to 0.85, but when we did that we received several errors in the logs. And the accuracy does not exceed 0%. 

We are starting from english pretrained model (en_PP-OCRv4_rec_train/best_accuracy.pdparams).

Dataset: 100k images on train dataset and 3k images on validation dataset

Regards,
Kevin

**Comment by GreatV** [comment]
Given the situation where your model's training accuracy is significantly higher than the validation accuracy, this is a classic case of overfitting. Here are some suggestions to improve your model's performance on the validation set:

- **Lower Learning Rate**: Since your validation accuracy is not improving, you might try reducing the learning rate even further. Overfitting often indicates that the model is learning too quickly, so slowing it down might help.
- Ensure that your validation set is representative of the different character lengths. If the validation set is too small or not diverse enough, it might not accurately reflect the model's performance.

**Comment by kevinzezel** [comment]
Hi,

Thank you!

We even tried with a very small learning rate like 0.0000002, but without success. 

What is more intriguing is that the 3 and 4 character base has the same font as the 7 character base and cannot perform well. In our last test, we took our 7-character base where we reached 95% acc in the validation base and truncated it to 3 characters to see what would happen. And started overfitting, reaching 98% training acc in the first few seasons. We were only able to make it work with 7 characters training dataset. 

It seems that recognizing a 3-character base is very easy for the model. Would that make sense? Is there any parameter that could be changed to reduce the size of the network?

**Reply by GreatV** [reply]
Since you are dealing with shorter text, ensure that `max_text_length` is appropriate. A smaller value can prevent the model from overfitting to unnecessary long sequences.

**Comment by kevinzezel** [comment]
Nice!

We will try. Thank you very much!
